hardware/kubernetes_client.go

```diff
@@ -0,0 +1,250 @@
+package hardware
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+	tinkv1alpha1 "github.com/tinkerbell/tink/pkg/apis/core/v1alpha1"
+	tink "github.com/tinkerbell/tink/pkg/controllers"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ Client = &KubernetesClient{}
+
+// KubernetesClient is a hardware client backed by a KubernetesClient cluster that contains hardware resources.
+type KubernetesClient struct {
+	client           ListerClient
+	close            func()
+	closeM           *sync.RWMutex
+	waitForCacheSync func(context.Context) bool
+}
+
+// NewKubernetesClientOrDie creates a new KubernetesClient client. It panics upon error.
+func NewKubernetesClientOrDie(config KubernetesClientConfig) *KubernetesClient {
+	client, err := NewKubernetesClient(config)
+	if err != nil {
+		panic(err)
+	}
+	return client
+}
+
+// NewKubernetesClient creates a new KubernetesClient client instance. It launches a goroutine to perform synchronization
+// between the cluster and internal caches. Consumers can wait for the initial sync using WaitForCachesync().
+// See k8s.io/client-go/tools/clientcmd for constructing *rest.Config objects.
+func NewKubernetesClient(config KubernetesClientConfig) (*KubernetesClient, error) {
+	opts := tink.GetServerOptions()
+	opts.Namespace = config.Namespace
+
+	// Use a manager from the tink project so we can take advantage of the indexes and caching it configures.
+	// Once started, we don't really need any of the manager capabilities hence we don't store it in the
+	// KubernetesClient
+	manager, err := tink.NewManager(config.Config, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	managerCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		if err := manager.Start(managerCtx); err != nil {
+			panic(err)
```

Comment:
If this err doesnt manifest until some set of circumstances that are only satisfied when an incoming request is handled, this might be difficult to troubleshoot. Having this called in func main somehow, might be something to consider.

Reply:
Now that I've plumbed in `oklog/run` I can do this. When it was originally possible this was nasty to do :(

Reply:
Scratch that. See my [other comment](https://github.com/tinkerbell/hegel/pull/82#discussion_r867362342).

Resolving this in favor of that discussion.